File: seminar_3/task_2/task2.cpp
#include <iostream>
#include <vector>
using namespace std;

void Shell_Sort(vector <int> &v){
    int gap = v.size() / 2;
    while(gap > 0){
        for(size_t current_position = gap; current_position < v.size(); current_position++){
            int m_gap = current_position;
            while(m_gap >= gap && v[m_gap] < v[m_gap - gap]){
                iter_swap(v.begin()+m_gap, v.begin() + m_gap - gap);
                m_gap-=gap;}
        }
        gap/=2;
    }
}

int feed_animals(vector <int>& animals, vector <int> &food){
    Shell_Sort(animals);
    Shell_Sort(food);
    size_t ans = 0;
    for(size_t i = 0; i < animals.size() && ans < animals.size(); i++)
        if(food[i] > animals[ans])
            ans++;
    return ans;
}

int main(void){
    vector <int> animals{1, 4, 3, 8}, food{8, 2, 3, 2};
    cout<<feed_animals(animals, food);
    return 0;
}

File: seminar_3/task_2/task_2.py
def quicksort(nums):
   if len(nums) <= 1:
       return nums
   else:
       q = nums[len(nums) // 2] 
       s_nums = [] # меньше q
       g_nums = [] # больше q
       e_nums = [] # равные q
       for n in nums:
           if n < q:
               s_nums.append(n)
           elif n > q:
               g_nums.append(n)
           else:
               e_nums.append(n)
       return quicksort(s_nums) + e_nums + quicksort(g_nums)


#задача 2 накормить животных

def feedAnimals(animals, food):
    if len(animals) == 0 or len(food) == 0:
        return 0
    quicksort(animals)
    quicksort(food)
    
    count = 0 
    for f in food:
        if f >= animals[count]:
            count += 1
        
        if count == len(animals):
            break
        
        
    return count    

animals = [2, 1, 4, 9]
food = [3, 2, 5, 7]
print(feedAnimals(animals, food))